cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md

```diff
@@ -901,6 +901,42 @@ void trivial_types_only(T t){
 }
 ```
 
+# Logging
+
+libcudf includes logging utilities (built on top of [spdlog](https://github.com/gabime/spdlog)
+library), which should be used to log important events (e.g. user warnings). This utility can also
+be used to log debug information, as long as the correct logging level is used. There are six macros
+that should be used for logging at different levels:
+
+* `CUDF_LOG_TRACE` - verbose debug messages (targeted at developers);
+* `CUDF_LOG_DEBUG` - debug messages (targeted at developers);
+* `CUDF_LOG_INFO` - information about rare events (e.g. once per run) that occur during normal
+execution;
+* `CUDF_LOG_WARN` - user warnings about potentially unexpected behavior, deprecations;
+* `CUDF_LOG_ERROR` - recoverable errors
+* `CUDF_LOG_CRITICAL` - unrecoverable errors (e.g. memory corruption)
+
+By default, messages below the `INFO` level are excluded from the log. In public builds, the code
+that logs at these levels is compiled out. This prevents logging of potentially sensitive data that
+might be done for debug purposes. Also, this allows developers to include expensive computation in
+the trace/debug logs, as the overhead will not be present in the public builds. The minimum enabled
+logging level is `INFO`, and it can be modified in multiple ways:
+
+* CMake configuration variable `LIBCUDF_LOGGING_LEVEL` - controls the minimum level of logging that
+will be compiled in the build.
+Available levels are `TRACE`, `DEBUG`, `INFO`, `WARN`, `ERROR`, `CRITICAL` and `OFF`.
+* Environment variable `LIBCUDF_LOGGING_LEVEL`, with the same values as the CMake variable.
+The difference is that disabled levels are present in the build, but the messages are excluded from
+the log.
+* Global logger object exposed via `cudf::logger()`. This API should not be used within libcudf to
+manipulate logging,
+it's purpose is to allow upstream users to configure libcudf logging to fit their application.
+
+The default log file is `cudf_log.txt` in the current working directory.
```

Comment:
I saw the default value is `INFO`. Do we expect text files to start showing up by default for all users of libcudf and/or Python cuDF? I am definitely not in favor of generating files without users opting in. stdout/stderr is fine.

Reply:
The current default level and sink type are based on RMM defaults. Upon further inspection, it looks like RMM does not have any INFO level messages, so only errors get logged into the file. 
Logging from multiple processes into a single file is not supported by spdlog, and libcudf is commonly used by multiple processes at the same time.
Given that creating files during normal execution is not desired, and the limitations of file logging, @bdice and I agreed on modifying the defaults to the following: 
Log warnings and more critical levels to stdout. Users can still redirect the log to a file of a given name using the environment variable.
Any feedback on this choice would be very welcome.

Reply:
Updated the default behavior to: log warnings and more critical levels to stdout

Reply:
Curious why choose stdout instead of stderr?  stdout is more likely than stderr to be used for something application specific.

Reply:
Did not make the initial choice carefully. Read up a bit today and changed to stderr.